content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,252 @@
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+## Introduction: the dangers of not checking function inputs
+
+R functions and R packages are a convenient way to share code with the rest of the world but it is generally not possible to know how, or with what precise aim in mind, others will use your code.
+For example, they might try to use it on objects that your function was not designed for.
+Let's imagine we have written a short function to compute the geometric mean:
+
+```{r}
+geometric_mean <- function(...) {
+  
+  return(prod(...)^(1/...length()))
+  
+}
+```
+
+When you tested the function yourself, everything seemed fine:
+
+```{r}
+geometric_mean(2, 8)
+
+geometric_mean(4, 1, 1/32)
+```
+
+But a different person using your function might expose it to situations it was not prepared to handle, resulting in cryptic errors or undefined behaviour:
+
+```{r, error = TRUE}
+# Input with factors instead of numerics
+geometric_mean(factor(2), 8)
+
+# Input with negative values
+geometric_mean(-1, 5)
+
+# Input with NAs
+geometric_mean(2, 8, NA)
+```
+
+Or worse, it could give an incorrect output:
+
+```{r}
+geometric_mean(c(2, 8))
+```
+
+Because of this, you need to make sure you return clear errors whenever your functions receives input it was not designed for.
+In this blog post, we review a range of approaches to help you check your function inputs and discuss some potential future developments.
+
+## Pre-requisite: thouroughly document your argument types
+
+You can notice from the simple example above that it's easy to pass invalid inputs to the `geometric_mean()` function because we didn't provide any documentation on what is or isn't a valid input.
+We won't go into details here but the [roxygen2](https://roxygen2.r-lib.org/) package provides a convenient way to generate documentation for R functions.
+Try to be as precise as possible when describing the required format for your inputs [^1].
+
+[^1]: [Some package developers even developed their own standardized way to document argument types and length](https://github.com/r-lib/withr/commit/42e503092046705f30032cb3a321d64b0e9383d4).
+    But there is currently no standard shared across the R community.
+
+```{r}
+#' @param name A character of length one with the name of the person to greet
+say_hello <- function(name) {
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+```
+
+Adding any kind of argument checking in the absence of good documentation would be vain and very frustrating for your users as they would have to figure out what is or isn't valid by trial and error.
+
+## Checking function inputs using base R
```

Comment:
Maybe also mention the rlang "equivalents"? https://rlang.r-lib.org/reference/index.html#check-arguments